### ISSUE TITLE
Bring up zed camera

### ISSUE THREAD

**Issue by darissa** [ROBOTIS-JAPAN-GIT/turtlebot3_slam_3d#17]
Hi @Affonso-Gui. I successfully run this code. Now, I wanna change to my environment which I am using ZED2 camera only  with yolov4 object detection. So first step I try to change the input to ZED camera, live input. FYI, I have changed several things:

1. changes node to /zed_node/left/image_rect_color
2. changes node to /zed_node/depth/depth_registered
3. changes node to /zed_node/left/camera_info
4. changes node to /zed_node/depth/camera_info
5. changes node to /zed_node/odom
6. changes node to /zed_node/point_cloud/fused_cloud_registered
7. comment out turtlebot3_robot_launch and publish_model.launch in turtlebot3_zed_bringup
8. comment out visual_odom line in turtlebot3_slam_3d.launch since already set to false
9. add nodelet depthimage_to_laserscan in rtabmap.launch
10. merge YoloObjectDetector.cpp file to this github https://github.com/AkellaSummerResearch/darknet_ros.git, because I already test, able to run live zed for yolo object detection in ROS.
11. add these lines in darknet_config.yaml

subscribers:

  camera_reading:
    topic: /zed_node/left/image_rect_color
    queue_size: 1

image_view:

  enable_opencv: true
  wait_key_delay: 1
  enable_console_output: true
  use_darknet: true

12. add these lines in yolov3.yaml

image_view:

  enable_opencv: true
  wait_key_delay: 1
  enable_console_output: true
  use_darknet: true

When I run roslaunch turtlebot3_slam_3d turtlebot3_slam_3d.launch, I got warning as below, and YOLO V3 window not pop-up.

/rtabmap/rtabmap: Did not receive data since 5 seconds! Make sure the input topics are published ("$ rostopic hz my_topic") and the timestamps in their header are set. If topics are coming from different computers, make sure the clocks of the computers are synchronized ("ntpdate"). If topics are not published at the same rate, you could increase "queue_size" parameter (current=600).
/rtabmap/rtabmap subscribed to (approx sync):
   /zed_node/odom \
   /zed_node/left/image_rect_color \
   /zed_node/depth/depth_registered \
   /zed_node/left/camera_info \
   /rtabmap/scan

Also, I checked all changed node are working except  /zed_node/point_cloud/fused_cloud_registered.

Already stuck with this thing about a week. Hope you can help. Thank you.

**Comment by darissa** [comment]
UPDATE:

Now, able to view YOLO V3 window and object detection works well. However, I got different error "segmented cloud size is zero" and map is not received in rviz. When I checked rostopic hz /map or any related topic to rtabmap, "no new messages" output.

**Comment by Affonso-Gui** [comment]
Hello,

This repository is quite old and active development has been dropped a few years ago when I changed position. In order to keep it running most of the external code has been pinpointed to a certain version (https://github.com/ROBOTIS-JAPAN-GIT/turtlebot3_slam_3d/blob/master/.rosinstall).
A lot has changed since then, unfortunately mostly without backward compatibility, as you have been experiencing.

This is probably a good time to update some of the code, do you think you can open a PR with these changes? Having access to your running code (and maybe a small bag file) would also help me a lot in debugging this problem!

**Comment by darissa** [comment]
@Affonso-Gui Sorry a bit late. I am new to github. So I just open a PR. You can checkout. I will  do the same for darknet_ros.git. Thank you so much for your support. I think this project is amazing, I actually seeking this kind of output quite some time. That's why I really try hard on your github. Tq Tq.

**Comment by Affonso-Gui** [comment]
Hello,

I couldn't find the Pull Request or any changes in your working branch... Could you check it once again and post the link to the new PR here?

**Comment by darissa** [comment]
Hi,

https://github.com/darissa/turtlebot3_slam_3d/tree/new_branch

**Comment by darissa** [comment]
This one for darknet_ros. I didn't change any for zed_wrapper.

https://github.com/darissa/darknet_ros_zed/tree/new_branch

Thank you

**Comment by darissa** [comment]
Hi,

I successfully generate the map input from live zed camera and svo file. However, i got warning:

 "[ WARN] (2022-04-14 13:23:10.245) SensorData.cpp:718::uncompressDataConst() Requested laser scan data, but the sensor data (193) doesn't have laser scan."  

and..even though yolo works well, i mean, they have the detection, but when run detection_collector.py, i got empty file.

![WhatsApp Image 2022-04-14 at 1 33 20 PM](https://user-images.githubusercontent.com/1271049/163320343-5b23682a-657f-4d51-8ddc-45a075186837.jpeg)

**Comment by Affonso-Gui** [comment]
> even though yolo works well, i mean, they have the detection, but when run detection_collector.py, i got empty file.

Something seems to be going wrong in the detection pipeline (which transforms 2d coordinates from the yolo detection image into 3d map coordinates). https://github.com/darissa/turtlebot3_slam_3d/blob/new_branch/launch/darknet.launch#L39-L60

 Which of the following topics are you receiving? (rostopic hz)

```
/darknet_ros/detection_image
/darknet_ros/label_image
/darknet_ros/cluster_points
/cluster_decomposer/centroid_pose_array
```

**Comment by Affonso-Gui** [comment]
> "[ WARN] (2022-04-14 13:23:10.245) SensorData.cpp:718::uncompressDataConst() Requested laser scan data, but the sensor data (193) doesn't have laser scan."

This warning seems a bit odd considering that you are not setting the `subscribe_scan` right now https://github.com/darissa/turtlebot3_slam_3d/blob/new_branch/launch/rtabmap.launch#L36

Have this occurred only once or several times? It might be simply your cache kicking in; try to explicitly set `subscribe_scan` to false instead of commenting it out.

**Comment by darissa** [comment]
Yes, you right. The collector doesn't work. Receiving these three only:

/darknet_ros/detection_image
/darknet_ros/label_image
/darknet_ros/cluster_points

Didn't received rostopic hz /cluster_decomposer/centroid_pose_array. I have checked, the point cloud topic received no new message. I subscribed to the topic zed_node/point_cloud/fused_cloud_registered

Yes, I already subscribe_scan to false. It occurred several times. I have checked the laserscan works but the warning still appear. I am using depthimage_to_laserscan node.

**Comment by darissa** [comment]
Thanks to you, now the collector is working after changing the topic to zed_node/point_cloud/cloud_registered. 

Then, I run the publisher, the label cannot publish. Now the warning 'segmented cloud  size is zero' and 'Requested laser scan data, but the sensor data (989) doesn't have laser scan' still have. And got another warning in rviz and the image cannot display. Also, as you can see in the figure, I think the z-axis is not at 0 right.  Where to fixed this parameter?

![WhatsApp Image 2022-04-16 at 7 00 51 AM](https://user-images.githubusercontent.com/1271049/163651676-1ca0b47e-8da4-4e14-b91d-cacd34abac7f.jpeg)

**Comment by Affonso-Gui** [comment]
Just to make sure, can you also see /zed_node/point_cloud/fused_cloud_registered ? What is the output from the following?
```
rostopic info /darknet_ros/cluster_points
rostopic info /zed_node/point_cloud/fused_cloud_registered
```
How many Hertz on each topic?

If everything seems fine and connected, it might be a synchronization issue. Even though we are setting `approximate_sync` to true, the default queue_size of 100 might still not be enough.

https://jsk-docs.readthedocs.io/projects/jsk_recognition/en/latest/jsk_pcl_ros/nodes/cluster_point_indices_decomposer.html

The best would be to have the same header in the yolo input and output image, and then set approximate_sync to false. This might have been covered here:
https://github.com/leggedrobotics/darknet_ros/issues/104
https://github.com/leggedrobotics/darknet_ros/pull/113
https://github.com/leggedrobotics/darknet_ros/issues/104#issuecomment-468262691


If the problem persists, uploading a short bag file in your fork would help me to dive deeper into the bug!

**Comment by Affonso-Gui** [comment]
Oh, nevermind, glad you fixed it! Can you also update your changes to the fork?

**Comment by darissa** [comment]
These samples are detected image. Is it weird to have -ve value? Is that why the cloud is in the middle of the plane of z-axis?

Found a tvmonitor at [-2.1779866218566895, -11.493218421936035, 0.6778092980384827]
Found a chair at [-3.11004638671875, -11.725285530090332, 0.34713223576545715]
Found a tvmonitor at [-2.2595298290252686, -8.947291374206543, 0.5882754921913147]
Found a person at [-2.6636433601379395, -9.517132759094238, 0.6968424916267395]
Found a chair at [-3.9244256019592285, -10.729484558105469, -0.09537818282842636]

**Comment by darissa** [comment]
> Oh, nevermind, glad you fixed it! Can you also update your changes to the fork?

Alright will do.

**Comment by Affonso-Gui** [comment]
> the label cannot publish

What are the contents of the detections_raw.db and detections_dbscan.db generated by the collector script?

**Comment by darissa** [comment]
These are the contents of the detection_dbscan.db

chair:
- [-2.880015383596006, -1.7301337926284126, -0.2924576758042626]
- [-1.6791381510821255, -1.762898781082847, -0.12793821654155513]
- [-2.5339683532714843, -5.661261749267578, -0.3524733191728592]
- [-1.8945059378941853, -9.098917961120605, -0.17952745345731577]
- [-1.4917239790613002, -13.167263594540684, -0.30208475240082905]
- [-3.5379942059516907, -13.63943362236023, -0.07198064960539341]
- [-2.8090590880467343, -11.47475895514855, -0.24177217254271874]
- [-3.8884234693315296, -11.116663773854574, -0.12134747670239045]
- [-4.635954101880391, -10.452462355295816, 0.07570578860274206]
- [-4.8179931640625, -7.491043651805205, -0.6048808045247022]
- [-5.883797109127045, -4.115401258071263, -0.7255580152074496]
- [-7.260846834916335, -0.09105827225945316, -1.045694770721289]
- [-6.58550877571106, -1.760263442993164, -0.7991519510746002]
- [-8.127901554107666, 3.3434295852979026, -1.4199368009964626]
- [-8.27373586851975, 5.204661303553088, -1.2015616112741931]
- [-12.937190554358743, 2.0198483995415946, -0.7772676010023464]
- [-10.045056343078613, -1.6525475978851318, -0.3837457001209259]
- [-8.10183596611023, -7.45281986395518, 0.10374711912785035]
- [-1.1352894766764208, -18.62890625, 0.8227511319247159]
- [-3.0469601829096957, -12.04646378643108, 0.40320187666506135]
keyboard:
- [-4.0618435236123895, -10.473235056950497, -0.18473029709779298]
- [-3.1297730207443237, -10.781843185424805, -0.2953672409057617]
microwave:
- [-9.890976905822754, 7.111268997192383, -0.5868287086486816]
person:
- [-0.12316341093211466, -1.7203101001936814, 0.11254604469085562]
- [-2.3055187463760376, -14.781752518245153, 0.8078382185527256]
refrigerator:
- [-5.639022668202718, -10.929157257080078, 0.21636213858922324]
tvmonitor:
- [-4.9266359329223635, -16.870916175842286, 1.0089244961738586]
- [-5.0904232774462015, -17.507807731628418, 0.9020655112607139]
- [-2.402776288986206, -14.107564926147461, 0.3127921879291534]
- [-4.29001404762268, -10.348020133972168, 0.02665666598593816]
- [-9.613080611595741, 2.117168527383071, 0.7752180053637578]
- [-10.616815376281739, 7.538045406341553, -0.05814264274667948]
- [-9.878599802652994, 7.231990814208984, -0.5984620849291483]
- [-10.921157836914062, 2.1259682178497314, -0.47452494502067566]
- [-4.600542978806929, -18.128211454911664, 2.462037519975142]

**Comment by darissa** [comment]
Hi @Affonso-Gui, already update my repo.

https://github.com/darissa/turtlebot3_slam_3d/tree/new_branch/launch.

**Comment by darissa** [comment]
@Affonso-Gui I just want to clarify this thing. I am using zed only. Then using depthtolaserscan node to have laser scan topic. But i got error 'Requested laser scan data, but the sensor data (989) doesn't have laser scan'. Is it because of this laser scan, it can't publish the label?

**Comment by Affonso-Gui** [comment]
@darissa Sorry for the delay, I have been busy lately but will take a look as soon as possible.

**Comment by ajrillah** [comment]
sorry, which file should i modify from zed wrapper for custom detector? because i am using ros2 zed wrapper and i want to use yolov8.
